Introduce tokens in web UI

eduardoj opened the pull request:
Manage user tokens from the web user interface.

This new option can  be accessed from "User Profile", "Manage Your Tokens".

This option will be shown only if the ~`user_profile_redesign`~ `trigger_workflow` feature flag is enabled.

This pull request only implements listing and deleting tokens.

Policies for using tokens from the web-ui were created under `app/policies/webui/tokens/`, to not colide with the existing token policies, under `app/policies/token/`.

-----------------
Switching to Draft. Things to be done before switching to "Ready for review":
* [x] Add pagination.
* [x] Add "Create token".
* [x] Don't show the token string field in the list of tokens.
-----------------

If this PR requires any particular action or consideration before deployment,
please check the reasons or add your own to the list:

* [ ] Requires a database migration that can cause downtime. Postpone deployment until maintenance window[1].
* [ ] Contains a data migration that can cause temporary inconsistency, so should be run at a specific point of time.
* [ ] Changes some configuration files (e.g. options.yml), so the changes have to be applied manually in the reference server.
* [ ] A new Feature Toggle[2] should be enabled in the reference server.
* [ ] Proper documentation or announcement has to be published upfront since the introduced changes can confuse the users.

[1] https://github.com/openSUSE/open-build-service/wiki/Deployment-of-build.opensuse.org#when-there-are-migrations
[2] https://github.com/openSUSE/open-build-service/wiki/Feature-Toggles-%28Flipper%29#you-want-real-people-to-test-your-feature

krauselukas commented on src/api/app/controllers/webui/users/tokens_controller.rb:
```diff
@@ -0,0 +1,24 @@
+  after_action :verify_policy_scoped, only: :index
+
+  def index
+    @tokens = policy_scope([:webui, Token])
```
```suggestion
    @tokens = policy_scope([:webui, Token]).page(params[:page])
```

krauselukas replied:
Could be a oneliner

eduardoj replied:
Done.

hennevogel commented on src/api/app/policies/webui/token/release_policy.rb:
```diff
@@ -0,0 +1 @@
+class Webui::Token::ReleasePolicy < Webui::TokenPolicy; end
```
Why namespace the policy?

eduardoj replied:
From the description of the PR: "Policies for using tokens from the web-ui where created under `app/policies/webui/tokens/`, to not colide with the existing token policies, under `app/policies/webui/tokens/`."

hennevogel replied:
I understand namespaces, but why do this? You expect them to behave differently then the API? And if so how?

vpereira commented:
Hey, I would like to raise some concerns regarding the logic being implemented. I'm not sure if they are already planned, but I would like to document those in my comment. 

![Screenshot_2021-07-26_16-30-35](https://user-images.githubusercontent.com/37418/127006961-1d8f3f5b-499c-48d1-8e3f-6e00ef13d1aa.png)

The column ID is unnecessarily long. Maybe we should introduce a name for the token. Looking my Gitlab/Github tokens, I like to name them, to know where I'm using them. It makes easier  to manage them, specially for long term tokens.

![Screenshot_2021-07-26_16-31-29](https://user-images.githubusercontent.com/37418/127007212-cd94c069-1eac-4e75-a5b8-1befbadbfa5d.png)

Today with `osc token`, I'm able to create a token for a project globally, without assign it to a specific package. I.e `osc token --create home:vpereirabr` works without a problem. However the WebUI is behaving differently here:

![Screenshot_2021-07-26_16-31-29](https://user-images.githubusercontent.com/37418/127007474-9db9c0ee-0fcd-4d01-8763-a2f0f8c5200d.png)

We are still missing the "edit" button, where we would be able for example for `workflow tokens` to update the SCM token without displaying it.. will that be implemented in this deliverable as spec'ed here https://trello.com/c/c1kcRa5n/1722-crud-ui-for-tokens?

eduardoj commented:
> The column ID is unnecessarily long.

This should be fixed since last week.

> Maybe we should introduce a name for the token. Looking my Gitlab/Github tokens, I like to name them, to know where I'm using them. It makes easier to manage them, specially for long term tokens.

It is also convinient to have a `created_at` field, to help to remember the tokens. We don't have the name field, neither the `created_at` field. In case of introducing those fields, I would prefer that to be done in another pull request.

> Today with `osc token`, I'm able to create a token for a project globally, without assign it to a specific package. I.e `osc token --create home:vpereirabr` works without a problem. However the WebUI is behaving differently here:

The error message you receive, explains it: "when providing an optional package, both project name and package name must be provided.". You only provided a project name. The same error happens if only a package name (without a project name) is provided.

> We are still missing the "edit" button, where we would be able for example for `workflow tokens` to update the SCM token without displaying it.. will that be implemented in this deliverable as spec'ed here https://trello.com/c/c1kcRa5n/1722-crud-ui-for-tokens?

I still miss it, too. I hope that it will be implemented in this deliverable. The edit button is not part of this pull request, so it will be easier to review. As soon as we have the list, create and delete functionality finished, I'm more than happy. Later on, we could add the "edit" button.

vpereira commented:
> The error message you receive, explains it: "when providing an optional package, both project name and package name must be provided.". You only provided a project name. The same error happens if only a package name (without a project name) is provided.

Ok sounds like if I run `osc token --create $PROJECT` the token get created but not assigned to any project, right? So you cannot have a token valid o all packages under a specific project?

vpereira commented:
> > The column ID is unnecessarily long.
> 
> This should be fixed since last week.

So maybe the review-app didn't redeploy it? https://obs-reviewlab.opensuse.org/eduardoj-webui_for_user_tokens/my/tokens

eduardoj commented:
> Ok sounds like if I run `osc token --create $PROJECT` the token get created but not assigned to any project, right?

Right.

> So you cannot have a token valid to all packages under a specific project?

Exactly, you cannot have a token valid to all packages under a specific project. These are the fields returned in the current table for tokens:
```
> Token
=> Token(id: integer, string: string, user_id: integer, package_id: integer, type: string, scm_token: string)
>
```
In case we want to provide that funcionality, we should add, at least, a `project_id` field to that database table.

dmarcoux commented on src/api/app/controllers/webui/users/tokens_controller.rb:
```diff
@@ -0,0 +1,84 @@
+  private
+
+  def set_token
+    @token = Token.find(params[:id])
```
I would move this code in `destroy` since it's only needed for this action. This would make the controller a tiny bit simpler with one less `before_action`.

eduardoj replied:
> I would move this code in `destroy` since it's only needed for this action. This would make the controller a tiny bit simpler with one less `before_action`.

In case we implement the update and trigger actions, it would be nice to have the `before_action` back.

dmarcoux replied:
If it's going to be used in some other upcoming actions, then fine :+1:.

dmarcoux commented on src/api/app/views/webui/users/tokens/index.html.haml:
```diff
@@ -0,0 +1,60 @@
+    .text-right
+      = link_to new_token_path do
+        %i.fas.fa-plus-circle.text-primary
+        Create Token
```
Comparing with other UIs, we have `Create (...)` link at the bottom left of lists. Examples:
- Create Package: https://obs-reviewlab.opensuse.org/eduardoj-webui_for_user_tokens/project/show/home:Admin
- Add Member: https://obs-reviewlab.opensuse.org/eduardoj-webui_for_user_tokens/groups/group_1

eduardoj replied:
Ok. I will move it to the bottom left of the list.

eduardoj replied:
Done.

dmarcoux commented on src/api/app/views/webui/users/tokens/index.html.haml:
```diff
@@ -0,0 +1,60 @@
+                  = link_to '#', title: 'Delete Token', class: 'text-nowrap',
+                    data: { toggle: 'modal', target: '#delete-token-modal', token_id: token.id, action: token_path(token) } do
+                    %i.fas.fa-times-circle.text-danger
+                    Delete
```
In tables like in [`Source Files` of a package](https://obs-reviewlab.opensuse.org/eduardoj-webui_for_user_tokens/package/show/home:Admin/ruby), we don't have text for `delete` icon.

```suggestion
```

eduardoj replied:
Applied.

dmarcoux commented on src/api/app/controllers/webui/users/tokens_controller.rb:
```diff
@@ -0,0 +1,84 @@
+  end
+
+  def validate_package
+    return unless @package_params[:project_name].present? ^ @package_params[:package_name].present?
```
Is this a typo or you really want to use `^` (binary XOR operator) here?

eduardoj replied:
I wanted to use the `^`  (binary XOR operator) here. For fetching a package we need a project name and a package name. Only a project name or only a package name is not enough to retrieve a package. See also [this comment](#issuecomment-886793814).

dmarcoux replied:
So having the more commonly used `&&` would also be fine here. Then this would become:

```suggestion
    return unless @package_params[:project_name].present? && @package_params[:package_name].present?
```

dmarcoux commented on src/api/app/policies/webui/token_policy.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def create?
+    # TODO: when user_profile_redesign is rolled out, remove the Flipper check
+    Flipper.enabled?(:user_profile_redesign, user)
```
We should also check if the token to be created is assigned to the user, no?  

```suggestion
    record.user == user && Flipper.enabled?(:user_profile_redesign, user)
```

eduardoj replied:
Fixed.

dmarcoux commented on src/api/app/controllers/webui/users/tokens_controller.rb:
```diff
@@ -0,0 +1,84 @@
+    end
+  end
+
+  def set_scm_token
```
This could go in the `initialize` method of `Token::Workflow` since it's only for that model.

dmarcoux commented:
@eduardoj @krauselukas As discussed in our programming session, I pushed 2 refactoring commits to the PR.

hennevogel commented:
Would it be possible to hide the project/package fields for the workflow token?

eduardoj commented:
> Would it be possible to hide the project/package fields for the workflow token?

Only for the workflow token... sure. On it.

hennevogel commented on src/api/app/views/webui/users/tokens/new.html.haml:
```diff
@@ -0,0 +1,84 @@
+              = render partial: 'webui/shared/autocomplete', locals: { html_id: 'token[project_name]', label: 'Project name', required: false,
+                                                                       data: { source: autocomplete_projects_path(local: true) } }
+            .col
+              = render partial: 'webui/shared/autocomplete', locals: { html_id: 'token[package_name]', label: 'Package name', required: false,
```
This field should be disabled until the project field is filled.

eduardoj replied:
Done.

eduardoj commented:
> > > The column ID is unnecessarily long.
> > 
> > This should be fixed since last week.
> 
> Maybe we didn't test it extensively on Google Chrome?

@vpereira: The `col-*` classes have been removed from the `td` elements. I found a reproducer with Google Chromium, and it is fixed now. Could you confirm that is working for you?

hennevogel commented on src/api/app/views/webui/users/tokens/new.html.haml:
```diff
@@ -0,0 +1,84 @@
+- operation_options = Token::OPERATIONS.map { |a| "Token::#{a}".constantize.token_name }
```
workflow only if you are in beta.

eduardoj replied:
I changed the flag checked from `user_profile_redesign` to `trigger_workflow`. I also updated the PR description. @hennevogel please review again.

eduardoj commented:
> > Would it be possible to hide the project/package fields for the workflow token?
> 
> Only for the workflow token... sure. On it.

@hennevogel: Fixed. Kudos to @krauselukas.

vpereira commented:
Hi @eduardoj 

I think we have some issues with the mobile resolution:

![Screenshot_2021-07-28_14-09-53](https://user-images.githubusercontent.com/37418/127320054-77259118-9f92-4add-a622-c827250b0675.png)

The search boxes beside each other makes hard to work with autocomplete, the boxes are quite small too

![Screenshot_2021-07-28_14-10-11](https://user-images.githubusercontent.com/37418/127320058-3470f44c-f8ef-452e-a9e0-2f62e62d0f13.png)

I think we do style the buttons differently on mobile no?

![Screenshot_2021-07-28_14-10-36](https://user-images.githubusercontent.com/37418/127320059-86bca2ae-220a-4154-8e5d-b80aa8e4bc8a.png)

The button "delete" is not usable in this resolution

eduardoj commented:
> I think we have some issues with the mobile resolution:

Out of curiosity, what viewport resolution are you using? Less than 320 pixels width?

> The search boxes beside each other makes hard to work with autocomplete, the boxes are quite small too

I will modify the form to wrap the input boxes in small width vieports.

> I think we do style the buttons differently on mobile no?

I suppose you're referring to their placement, not to show them without space between them. I'll fix that. Please correct me if you wanted to say something different.

> The button "delete" is not usable in this resolution

I will add padding to them.

vpereira commented:
> Out of curiosity, what viewport resolution are you using? Less than 320 pixels width?

I'm using Iphone 5S

eduardoj commented:
> > The search boxes beside each other makes hard to work with autocomplete, the boxes are quite small too
> 
> I will modify the form to wrap the input boxes in small width vieports.

Done.

> > I think we do style the buttons differently on mobile no?
> 
> I suppose you're referring to their placement, not to show them without space between them. I'll fix that. Please correct me if you wanted to say something different.

The placement and distance between buttons has been adjusted. See this point and the previous one in this screenshot:

![Screenshot from 2021-07-29 08-40-36](https://user-images.githubusercontent.com/24919/127444169-b2dfdd43-b24c-42e4-86bc-471c6bdd3f79.png)

> > The button "delete" is not usable in this resolution
> 
> I will add padding to them.

I added a commit to add button classes to the icon. So it will be easier to click, also in mobile devices.

I also added a class to wrap long package names:

![Screenshot from 2021-07-29 08-37-46](https://user-images.githubusercontent.com/24919/127443708-bb7b4880-5b2a-4e82-9f8d-fd765c8fc669.png)

vpereira reviewed (approved):
approved. Maybe would be nice to squash some commits?

hennevogel commented on src/api/app/models/token.rb:
```diff
@@ -41,6 +47,30 @@ def package_find_options
+
+  protected
+
+  def validate_package
```
This part is not as simple as that. Each token type has it's own way of finding a package with various options and quirks. Including which names it accepts etc.

eduardoj replied:
I moved back the validation of the package parameters (project_name and package_name) and its search to the controller.

hennevogel commented on src/api/app/views/webui/users/tokens/_actions.html.haml:
```diff
@@ -0,0 +1 @@
+= link_to('Back to List of Tokens', tokens_path, title: 'Back to List of Tokens', class: 'btn btn-outline-secondary px-4 mt-3 mt-sm-0')
```
💭  Browsers have back buttons, is this really needed?

eduardoj replied:
In this case I would say yes. This 'Back to List of Tokens' link will be shown after a token has been created. If you press the back button of your browser, you will not request a page in the server, but you will still see the previous list of tokens, without the newly created one. It doesn't refresh automatically.

Do you have some other solution in mind?

eduardoj commented:
> Maybe would be nice to squash some commits?

I'm up for it. What of them, exactly, if I may ask?

hennevogel commented on src/api/spec/controllers/webui/users/tokens_controller_spec.rb:
```diff
@@ -0,0 +1,169 @@
+        get :index
+      end
+
+      it { expect(response).to render_template('webui/users/tokens/index') }
```
What would stop this from functioning? :)

eduardoj replied:
Lines `it { expect(response).to render_template('...') }` removed. As this was the only thing that was tested for tne `new` action, that test was also removed.

hennevogel commented on src/api/spec/controllers/webui/users/tokens_controller_spec.rb:
```diff
@@ -0,0 +1,169 @@
+
+  describe 'GET #index' do
+    context 'user is not logged in' do
+      it_behaves_like 'require logged in user' do
```
`require_login` is already tested in the `WebuiController` spec. No need to do it again here for every action.

eduardoj replied:
"require logged in user" tests removed.

vpereira commented:
> I'm up for it. What of them, exactly, if I may ask?

I don't know, If all the introduced changes just work together, 1 commit: "Introduce tokens in web UI", would be enough. The little single commits are good to have in the PR, but they shouldn't land in master as separated unities. Like you added first the index, then you added the button for create a new token, then you added a button to delete.. those should be just one commit, IMO.

hennevogel commented on src/api/spec/controllers/webui/users/tokens_controller_spec.rb:
```diff
@@ -0,0 +1,169 @@
+      context 'no project and package parameters' do
+        let(:form_parameters) { { token: { operation: 'release' } } }
+
+        include_examples 'check for flashing a success'
```
I would feel more comfortable if you would test that the things you want to change, change. Instead of testing this only from the "outside" via the flash.

eduardoj replied:
I added tests for checking the number of tokens. Please have a look again.

eduardoj commented on src/api/app/models/token.rb:
```diff
@@ -3,15 +3,21 @@ class Token < ApplicationRecord
+  validates :scm_token, absence: true, if: -> { type != 'Token::Workflow' }
+  validate :validate_package
+  # TODO: validate_package: Is this for all tokens except Token::Workflow?
+  # TODO: validate_package: Check if this is affecting the API or maybe even osc
```
Let's do it now.

eduardoj replied:
After removing `validate_package` from the Token model, both TODOs can be removed.

hennevogel commented on src/api/app/models/token.rb:
```diff
@@ -3,15 +3,18 @@ class Token < ApplicationRecord
   belongs_to :package, inverse_of: :tokens
 
-  attr_accessor :object_to_authorize
+  attr_accessor :object_to_authorize, :operation, :project_name, :package_name
```
No need for those anymore right?

eduardoj replied:
Nop. Removed.
